feat: add initial NATS setup

jonaslagoni opened the pull request:
This PR adds the initial support for NATS in the most basic form.

Do let me know if anything is missing or needs to change 🙂 

Todo:
- [ ] Do a final test

jonaslagoni reviewed (commented):
What is the best way to check everything works as intended, or which one do you use @lbroudoux?

lbroudoux commented:
Wouah! That's awesome! TBH whole chain testing is done manually at the moment... never found enough time to seriously tackle its automation.

lbroudoux commented:
As I have just started working on 1.7 that is Java 17, Spring Boot 2.7.3 and Quarkus 2 based, would you mind switching this PR to 1.7.x branch? Apart the Java version that you have to be installed on your machine it shouldn't break anything in the code.

jonaslagoni commented:
My pleasure 🙂 Thanks for the quick reply!

> TBH whole chain testing is done manually at the moment... never found enough time to seriously tackle its automation.

Alright, I am gonna take a look at some mocking setups and see if I cant write some basic tests for least the consumption and producer 🤔 Otherwise I will just do a manual check.

> As I have just started working on 1.7 that is Java 17, Spring Boot 2.7.3 and Quarkus 2 based, would you mind switching this PR to 1.7.x branch? Apart the Java version that you have to be installed on your machine it shouldn't break anything in the code.

Definitely 👍

lbroudoux commented on install/kubernetes/README.md:
```diff
@@ -235,6 +235,14 @@ Here are below the configuration properties of the MQTT support feature:
+
+| Section               | Property   | Description                                                                                                                              |
+| --------------------- | ---------- | ---------------------------------------------------------------------------------------------------------------------------------------- |
+| `features.async.nats` | `url`      | **Optional**. The URL of NATS broker (eg: `my-nats-broker.example.com:4222`). Default is undefined which means that feature is disabled. |
```
Do you plan to manage authentication to NATS broker as well? Ideally that should be mandatory for release

jonaslagoni replied:
For the first PR I actually did not think about doing it, for the simple reason I feel the PR is large enough. I know NATS has a bunch of different authentication options, so it might need some discussion on how to handle some of it 😄 So it's just my personal preference to slowly add features and don't do a hail-mary one massive PR.

jonaslagoni replied:
Are there any specifics as to which features you see as a must-have for it to be merged?

lbroudoux replied:
Totally agree on the approach of incremental PR! Authentication is a must have to pretend "Enterprise Ready" but I have no list on required options... I think it totally depends on what's commonly used on NATS side. We'll see ;-)

lbroudoux commented on minions/async/src/main/java/io/github/microcks/minion/async/consumer/AMQPMessageConsumptionTask.java:
```diff
@@ -168,7 +168,7 @@ public void handleDelivery(String consumerTag,
                      throws IOException {
                   logger.info("Received a new AMQP Message: " + new String(body));
-                  // Build a ConsumedMessage from MQTT message.
+                  // Build a ConsumedMessage from AMQP message.
```
Thanks for the typo fix ;-)

lbroudoux commented on minions/async/src/main/java/io/github/microcks/minion/async/consumer/NATSMessageConsumptionTask.java:
```diff
@@ -0,0 +1,155 @@
+            // Build a ConsumedMessage from NATS message.
+            ConsumedMessage message = new ConsumedMessage();
+            message.setReceivedAt(System.currentTimeMillis());
+            message.setPayload(msg.getData());
```
Is NATS message having some headers we should also consider here?

jonaslagoni replied:
Yes, there definitely is, added

lbroudoux commented on minions/async/src/main/resources/application.properties:
```diff
@@ -46,6 +49,6 @@ amqp.username=microcks
 
 # Configure the minion own behavioral properties.
-minion.supported-bindings=KAFKA,WS
+minion.supported-bindings=KAFKA,WS,NATS
```
Can you please remove NATS from the defaults so that we keep a small default setup?

lbroudoux commented on CONTRIBUTING.md:
```diff
@@ -56,9 +56,13 @@ $ cd webapp
 [...]
 $ cd ../install/docker-compose
-$ docker-compose -f microcks-mongodb.yml up -d
+$ docker-compose -f docker-compose.yml up -d
```
Thanks for the fix!

lbroudoux reviewed (changes requested):
Awesome! Thanks again!
This is an initial review some very small changes requests. Looks very good to me!

jonaslagoni commented on minions/async/src/main/java/io/github/microcks/minion/async/producer/NATSProducerManager.java:
```diff
@@ -0,0 +1,119 @@
+    * @param topic The destination topic for message
+    * @param value The message payload
+    */
+   public void publishMessage(String topic, String value) {
```
Added 👍

jonaslagoni commented:
I still need to test through it, but applied all your comments @lbroudoux 🙂

lbroudoux commented on minions/async/src/main/java/io/github/microcks/minion/async/consumer/NATSMessageConsumptionTask.java:
```diff
@@ -0,0 +1,175 @@
+     * The string for Regular Expression that helps validating acceptable
+     * endpoints.
+     */
+    public static final String ENDPOINT_PATTERN_STRING = "nats://(?<brokerUrl>[^:]+(:\\d+)?)";
```
I'm surprised you don't need a topic, queue, destination or subject name here so that you can later listen on specified subject ... ?

jonaslagoni replied:
I am definitely missing the connection of how consumption task can get the topic to listen to...

But where does that come from 🤔 I see the other protocols get this through the endpoint pattern string. Is that a microcks thing? Cause in general you cannot provide topic through the connection endpoint for NATS 🤔

lbroudoux replied:
Yes, this a pure Microcks thing : it's up to you to decide which elements you'll need in an endpoint URL so that you'll be able to later use them for connecting the broker with required options in the consumption task. Sometimes, I get inspiration from how Camel endpoints are expressed, which options are provided, ... Check the [Camel for NATS endpoint](https://camel.apache.org/components/3.18.x/nats-component.html#_endpoint_options).

lbroudoux commented on minions/async/src/main/java/io/github/microcks/minion/async/consumer/NATSMessageConsumptionTask.java:
```diff
@@ -0,0 +1,175 @@
+            messages.add(message);
+        });
+
+        d.subscribe("subject");
```
I think it should be replaced with a subject name that would be acquired from parsing the endpoint url.

jonaslagoni replied:
Yep 👍

lbroudoux reviewed (commented):
Thanks a lot for latest additions! Let me know when you think that it could be ready for a test ride 😉

jonaslagoni commented:
Alright, made the following changes:

- Indirectly through the connection, it supports basic authorization through [username/password](https://docs.nats.io/using-nats/developer/connecting/userpass) and [token](https://docs.nats.io/using-nats/developer/connecting/token).
- Fixed that NATS has multiple protocols supports that the regex did not match (TCP, nats, and WS)
- Fixed that the consumption task did not include the topic in the test

Here is where I am currently unable to progress any further in my testing:

1. I am failing to build the async minion docker image as it seems the documentation is slightly outdated as its fairly different from the main webapp (`microcks\minions\async\src\main\docker`) as it does not have the same scripts to build it, and when manually doing it it creates trouble:
    - Trying to build the native container ` docker build -f src/main/docker/Dockerfile.native -t microcks/microcks-async-minion .` fails with the following:
```
[+] Building 1.2s (8/8) FINISHED
 => [internal] load build definition from Dockerfile.native                                                        0.0s
 => => transferring dockerfile: 39B                                                                                0.0s
 => [internal] load .dockerignore                                                                                  0.0s
 => => transferring context: 2B                                                                                    0.0s
 => [internal] load metadata for registry.access.redhat.com/ubi8/ubi-minimal:8.4-212                               0.6s
 => [internal] load build context                                                                                  0.0s
 => => transferring context: 28B                                                                                   0.0s
 => [1/4] FROM registry.access.redhat.com/ubi8/ubi-minimal:8.4-212@sha256:c536d4c63253318fdfc1db499f8f4bb0881db7f  0.0s
 => CACHED [2/4] WORKDIR /work/                                                                                    0.0s
 => CACHED [3/4] COPY --chown=1001:root target/*-runner /work/application                                          0.0s
 => ERROR [4/4] RUN chmod 775 /work /work/application   && chown -R 1001 /work   && chmod -R "g+rwX" /work   && c  0.5s
------
 > [4/4] RUN chmod 775 /work /work/application   && chown -R 1001 /work   && chmod -R "g+rwX" /work   && chown -R 1001:root /work:
#8 0.440 chmod: cannot access '/work/application': No such file or directory
------
executor failed running [/bin/sh -c chmod 775 /work /work/application   && chown -R 1001 /work   && chmod -R "g+rwX" /work   && chown -R 1001:root /work]: exit code: 1
```  
2. I think I either miss some implementation or simply cant operate the microcks app correctly as it keeps saying kafka regardless of what I provide it
![image](https://user-images.githubusercontent.com/13396189/206868564-b65a1446-c25e-4a7d-8e0d-614076f2e4a5.png) 
3. Is there a reason why I need to use the endpoint when running the tests from an AsyncAPI document? Seems to me it would be far better to select an operation, as you already do (that already contain the topic), and you could select the server to perform the test on. Or is there some use-case I don't see where this is useful? 


I might need some guidance @lbroudoux for solving 1 and 2 😅

lbroudoux commented:
Hey Jonas, 

Thanks a lot for the changes you've pushed. Here are below answers to your questions.

> 1. I am failing to build the async minion docker image as it seems the documentation is slightly outdated as its fairly different from the main webapp (`microcks\minions\async\src\main\docker`) as it does not have the same scripts to build it, and when manually doing it it creates trouble:

At the moment, we're not using the native build for Quarkus but the one based on JVM. You may use `docker build -f src/main/docker/Dockerfile.jvm -t microcks/microcks-async-minion .` to build the image.
 
> 2. I think I either miss some implementation or simply cant operate the microcks app correctly as it keeps saying kafka regardless of what I provide it

Kafka binding is the default one if no other is provided. Have you specified a NATS binding in your AsyncAPI file? Is it correctly parsed by the `AsyncAPIImporter` class?

> 3. Is there a reason why I need to use the endpoint when running the tests from an AsyncAPI document? Seems to me it would be far better to select an operation, as you already do (that already contain the topic), and you could select the server to perform the test on. Or is there some use-case I don't see where this is useful?

I was wondering the same things at first but I realized that:
* All the servers used by an organization are not necessarily specified in the AsyncAPI file,
* The `channel` to `topic`, `queue`, `destination` or whatever translation is not that straightforward depending on the procotol (as an example `/` is a forbidden characters in Kafka topics and some people may translate it in `-` and others in `_`)

That's why I ended up with full endpoint syntax that seems more flexible. In a future release, I'd like to introduce some kind of environment management that would allow to abstract and not having to repeat this endpoint definition.

lbroudoux commented:
I've got some free cycles this week and I'd like to merge this one even if still marked as Draft. I think I'd be able to make the final adjustments to make it work.  What could be very useful - if you're ok with it - is the contribution of testing clients (consumer and producer) for NATS that could live in our `api-tooling` repo. This is where I have my reference scripts for running integration tests and demo on various protocol. See [here](https://github.com/microcks/api-tooling/tree/main/async-clients).

What do you think?

jonaslagoni commented:
> At the moment, we're not using the native build for Quarkus but the one based on JVM. You may use docker build -f src/main/docker/Dockerfile.jvm -t microcks/microcks-async-minion . to build the image.

Ahh alright!

> Kafka binding is the default one if no other is provided. Have you specified a NATS binding in your AsyncAPI file? Is it correctly parsed by the AsyncAPIImporter class?

Yep, I tried to use the following:

```
{
    ...
    "channels": {
        "v0/rust/servers/{server_id}/events/started": {
            "description": "Channel for the API to process for when a server has started",
            "parameters": {
                "server_id": {
                    "description": "The ID of the server",
                    "schema": {
                        "type": "string"
                    }
                }
            },
            "subscribe": {
                "operationId": "ServerStarted",
                "description": "The Rust server can publish to this channel when the server has started",
                "message": {
                    "name": "ServerStarted",
                    "payload": {
                        "$schema": "http://json-schema.org/draft-07/schema",
                        "type": "object",
                        "$id": "ServerStarted",
                        "required": ["timestamp"],
                        "properties": {
                            "timestamp": {
                                "type": "string",
                                "example": "2016-08-29T09:12:33.001Z"
                            }
                        }
                    },
                    "examples": [
                        {
                            "name": "laurent",
                            "summary": "Example for Laurent user",
                            "payload": {"timestamp": "2016-08-29T09:12:33.001Z"}
                        }
                    ],
                    "bindings": {
                        "nats": {
                            "bindingVersion": "0.1.0"
                        }
                    }
                },
                "bindings": {
                    "nats": {
                        "bindingVersion": "0.1.0"
                    }
                }
            },
            "bindings": {
                "nats": {
                    "bindingVersion": "0.1.0"
                }
            }
        },
    }
    ...
}
```

>That's why I ended up with full endpoint syntax that seems more flexible. In a future release, I'd like to introduce some kind of environment management that would allow to abstract and not having to repeat this endpoint definition.

Makes sense 👍 

In 3.0, the [introduction of the `address` keyword in 3.0.0](https://github.com/asyncapi/spec/pull/827/files#r964379549) ensures that you can use it entirely as is going forward 🙂 

But yea, definitely correct regarding the servers 👍 

> I've got some free cycles this week and I'd like to merge this one even if still marked as Draft

That's entirely up to you 🙂
